### PR TITLE
RactiveModel: do not set if unset has been called

### PR DESCRIPTION
**Pull request by kaicataldo**
Part of this fix: https://github.com/behance/poet/pull/342

### CODE REVIEW

File: test/specs/ractive/Model.js [hunk 1]
```diff
@@ -58,6 +58,21 @@ define([
...
 
+    describe('#set', function() {
+      it('should set item to ractive model basd on the path', function() {
+        expect(ractive.get('pages./page-two.title')).toBe('Page Two');
+        ractiveModel.set('!pages./page-two.title', 'Page 2');
+        expect(ractive.get('pages./page-two.title')).toBe('Page 2');
+      });
+
+      it('should not set if ractive model has already been unset', function() {
+        ractiveModel.unset('!pages./page-one');
```
Comment (mikesherov) [inline test/specs/ractive/Model.js]: I don't get what this parameter does?

Comment (mrjoelkemp) [inline test/specs/ractive/Model.js]: This is `page-one` while the other tests are `page-two`. I'm a little confused.

File: test/specs/ractive/Model.js [hunk 2]
```diff
@@ -58,6 +58,21 @@ define([
       });
     });
 
+    describe('#set', function() {
+      it('should set item to ractive model basd on the path', function() {
+        expect(ractive.get('pages./page-two.title')).toBe('Page Two');
+        ractiveModel.set('!pages./page-two.title', 'Page 2');
```
Comment (mikesherov) [inline test/specs/ractive/Model.js]: Remove `!`

File: test/specs/ractive/Model.js (hunk 1 above)
Comment (kaicataldo) [inline test/specs/ractive/Model.js]: Sorry, this is just plain wrong. Removed the param.

File: test/specs/ractive/Model.js [hunk 3]
```diff
@@ -58,6 +58,21 @@ define([
...
+        expect(ractive.get('pages./page-two.title')).toBe('Page 2');
+      });
+
+      it('should not set if ractive model has already been unset', function() {
```
Comment (mikesherov) [inline test/specs/ractive/Model.js]: Need to think on this test more: if you get *after* unset, it needs to return `undefined`. But that doesn't prove that your `set` behavior works...

File: ractive/Model.js [hunk 4]
```diff
@@ -30,18 +30,22 @@ define([
 
     /**
      * @param  {String} prop
-     * @return {*}
+     * @return {*|null}
```
Comment (kaicataldo) [inline ractive/Model.js]: This is something I've come across a few times - is this the correct JSDoc notation for this? I never know whether it should be `undefined`, `null`, and I've also seen `void`

File: test/specs/ractive/Model.js (hunk 3 above)
Comment (mrjoelkemp) [inline test/specs/ractive/Model.js]: I don't follow what we're trying to achieve. Unsetting a keypath nullifies the value. It shouldn't render the ractiveModel unusable.

It sounds like you want to destroy the ractiveModel instance at the consumer level to render it useless on an async set of the ractiveModel

File: ractive/Model.js (hunk 4 above)
Comment (mrjoelkemp) [inline ractive/Model.js]: What does the jsdoc documentation say?

Comment (kaicataldo) [inline ractive/Model.js]: Maybe it's right under my nose and I'm missing it, but I actually have spent time trying to Google it/look through the docs. These are the closest things I've found: 
 - http://usejsdoc.org/tags-returns.html
 - http://usejsdoc.org/tags-type.html

Apologies if it's obvious and I'm just missing it.

File: test/specs/ractive/Model.js (hunk 3 above)
Comment (kaicataldo) [inline test/specs/ractive/Model.js]: I'm gleaning info as I go - feel like my understanding keeps changing drastically every time I go make a change...The change in this PR doesn't really make sense to me now either. It now returns `undefined` for ALL calls to `set` or `get` if `unset` is called on any keypath...

File: ractive/Model.js (hunk 4 above)
Comment (mrjoelkemp) [inline ractive/Model.js]: Not sure. `?*` (not sure if that's valid) or `{}` makes sense to me. 

image

File: test/specs/ractive/Model.js (hunk 3 above)
Comment (mikesherov) [inline test/specs/ractive/Model.js]: conside a ractiveModel around an array:

```
var ractive = new Ractive({
  data: {
    foo: [
      { id: 1 },
      { id: 2 }
    ]
  }
});

var ractiveModel = new RactiveModel('foo.0', ractive);
ractiveModel.set('beep', 'boop');
ractiveModel.unset();
ractiveModel.set('shmeep', 'shmoop');
```

You'd expect data to be: ` [{ id: 2 }]` but instead it's `[{ id: 2, shmeep: 'shmoop' }]`.

Comment (mrjoelkemp) [inline test/specs/ractive/Model.js]: That makes sense. Great example

Comment (kaicataldo) [inline test/specs/ractive/Model.js]: As discussed with @mikesherov in Slack, this current solution does not work in the case of a consumer unsetting property lower down in the tree, as it makes the properties higher up inaccessible.

**Pull request closed without merging**